### PR TITLE
Refactor `mouse_entered` and `mouse_exited` signals

### LINKED ISSUE #20881: "mouse_entered" and "mouse_exited" signals are broken while holding down a mouse button
**Godot version:**
fbb5ca4

**OS/device including version:**
Linux Manjaro

**Issue description:**
The "mouse_entered" and "mouse_exited" signals of control nodes don't get fired while holding any mouse button down when the click (any mouse button down) started on a control that does not ignore the mouse. 

it's actually because in viewport.cpp in line 1843 to 1850 "over" get's set to the mouse_focus
```
if (gui.mouse_focus) {
	over = gui.mouse_focus;
	//recompute focus_inv_xform again here
} else {
	over = _gui_find_control(mpos);
}
```

and shortly after (  line 1884 to 1893 ) the mouse_exited and mouse_entered signals get not fired because the gui.mouse_focus and the gui.mouse_over are the same

```
if (over != gui.mouse_over) {

	if (gui.mouse_over)
		gui.mouse_over->notification(Control::NOTIFICATION_MOUSE_EXIT);

	_gui_cancel_tooltip();

	if (over)
		over->notification(Control::NOTIFICATION_MOUSE_ENTER);
}
```

This may be a desired behavior, for a lot of cases.
But there are also cases where it is very useful to send those signals. (I'll update this post soon to show such a case)

**Steps to reproduce:**
1. Connect the "mouse_entered" or "mouse_exited" signal of any control node
2. add another control node, that does not ignore the mouse
3. start your drag on the control element of point 2 and drag over the control element from point 1
4. notice that no "mouse_entered" and no "mouse_exited" signals get fired

### LINKED ISSUE #66293: Embedded window doesn't emit mouse_exited / mouse_entered.
### Godot version

4.0.beta1

### System information

Windows 10

### Issue description

Embedded window doesn't emit `Window.mouse_exited` or `Window.mouse_entered` when the player move the mouse into or out of the window's area. (Embedded window meaning its parent viewport has `gui_embed_subwindows` = false.) 

Non-embedded windows will correctly emit both signals.

Also, both the embedded window and its parent will emit `Window.window_input(event)`, as if they're a same window - which is technically right on the OS level. So it might not be a bug.

If it's not a bug, how do we check if the mouse exited an embedded window?

### Steps to reproduce

See reproduction project.

### Minimal reproduction project

[mouse_exited.zip](https://github.com/godotengine/godot/files/9632710/mouse_exited.zip)

### PR DESCRIPTION
**Pull request by Sauermann**
fix variant of #20881 where Controls are in different Viewports
fix #66293
fix no longer #57349 (that is already fixed for the case of a single Viewport)
prerequisite for #67531
prerequisite for #77781
addresses #48634 after #77933
related to #54565 (maybe supersedes it)
done: don't merge before #67903 (fix display-server differences between Windows and linuxbsd)

The current implementation for signals `mouse_entered` and `mouse_exited` have shortcomings that relate to focused windows and pressed mouse buttons. For example a `Control` can be hovered by mouse, even if it is occluded by an embedded `Window`.

This patch changes the behavior, so that `Control` and `Viewports` send their mouse-enter/exit-notifications based solely on mouse position, visible area and input restrictions and not on which window has focus or which mouse buttons are pressed. This implicitly also changes when the `mouse_entered` and `mouse_exited` signals are sent.

Rewriting a core functionality like `mouse_entered` and `mouse_exited` should be tested thoroughly. While I tested it in lots of situations and had a look at the locations where it is used in the code base, it can be, that I have missed something.

So far, tests on the following platforms were successful:
- linuxbsd
- Windows
- Web (Firefox, Chromium, Edge)

Other platforms:
- macOS: not yet tested
- Android / iOS: Not sure, if the concept of `mouse_entered` and `mouse_exited` makes sense in this context.
- Android: Signals do not trigger with touch and drag

The concept for this implementation is:
- When a mouse-event is sent from the DisplayServer to a Window, it first recursively determines mouse-over states for all Viewports and sends mouse-enter/exit notifications and signals.
- Afterwards the event handling happens just as before, without the enter/exit-notifications.
- Additional adjustments to how mouse-over-state is dropped were necessary.

MRP: [BugVisualMouseNotifications.zip](https://github.com/godotengine/godot/files/9865023/BugVisualMouseNotifications.zip) (updated 2022-10-26)

This functionality can not be implemented as a part of `Viewport::_gui_input_event`, because of its interplay with Windows and because `Viewport::_gui_input_event` is based on input and not on visibility.

https://user-images.githubusercontent.com/6299227/197390115-33c8b18d-69ff-4d87-a293-8378fe099db9.mp4

Updated 2022-11-13: Fix location of `sw->_update_mouse_over` call.
Updated 2022-11-15: Fix merge conflict with #68019
Updated 2022-11-19: Fix location of `sw->_update_mouse_over` call.
Updated 2023-01-06: Fix merge conflict with #69972
Updated 2023-01-27: Fix merge conflict with #66692
Updated 2023-02-18: Fix merge conflict with #72764
Updated 2023-06-07: Fix merge conflict with #77933
Updated 2023-06-29: Made necessary adjustments for Gui in 3D Demo
Updated 2023-07-07: Fix merge conflict with #78078
Updated 2023-07-27: Fix merge conflict with #79248

### CODE REVIEW

**Comment by lmurray** [comment]
I disagree with adding new signals that differ from the existing ones as a single mouse cursor can only be in one location at a time so it cannot be both inside and outside of a region at the same time which is what is implied by having multiple signals. No other UI toolkit that I know has two separate entered/exited signals that have different rules. Instead I think the existing signals are incorrectly implemented and need to be modified to correctly detect when the cursor enters and exits a region under all circumstances.

**Comment by Sauermann** [comment]
@lmurray I also would prefer to replace the old signals. However i have not yet tested the places, where the old signals are used within the code-base. So I provided new signals in this initial version of this PR, so that they can be tested and compared.

**Review by bruvzg — COMMENTED** [review]
Adding a new signals/notifications is probably unnecessary, it should replace existing viewport signals/methods `NOTIFICATION_VP_MOUSE_*` (for native windows there's `NOTIFICATION_WM_MOUSE_*`).

**Comment by Sauermann** [comment]
I have updated the PR, so that now `Viewports` emit `NOTIFICATION_VP_MOUSE_*` notifications based on the new method.
The behavior changes, but I didn't experience any problematic effects.

I left `NOTIFICATION_WM_MOUSE_*` alone, because this works already in the expected way.

My next step will be investigating, if replacing the `Control`-notifications by the new method would introduce unwanted behavior.

**Comment by Sauermann** [comment]
With the latest update, I have implemented a replacement for the `Control`-signals `mouse_enter/exit`. Feedback about how this works on other platforms than Linux would be appreciated.

**Comment by Sauermann** [comment]
Since the Windows and linuxbsd Display server behave quite differently, it makes sense to merge this PR only after the differences are solved. The PR #67903 makes this happen.

**Comment by KoBeWi** [comment]
I gave it a test and found a bug very quickly:
mycha_yfOC70IBoB
The cursor shape misbehaves when going from connection dialog to script editor. Without this PR it consistently uses caret shape.

**Comment by Sauermann** [comment]
Thanks for the feedback @KoBeWi 
This issue was caused by the Windows Display Server which did not send a `WINDOW_EVENT_MOUSE_EXIT` event in the case that a window got closed.

File: doc/classes/Node.xml
Comment (KoBeWi) [inline doc/classes/Node.xml:930]: Can a Control obscure a Viewport?
btw you seem to use `[]` inconsistently compared to descriptions above.

Comment (Sauermann) [reply doc/classes/Node.xml:930]: With the latest update, I use [] consistently in the docs.

**Comment by bruvzg** [comment]
Seems to break popup `y` positions on macOS (but it might be unrelated and just need a rebase, I'll retest later). Couldn't find any other bugs, mouse enter/exit seems to work ok.

Screenshot 2023-01-26 at 13 13 14

**Comment by Sauermann** [comment]
I'm not sure, if this should be moved to 4.1.
It has potential for regressions.
@bruvzg have you had any chance to retest, if your problem was actually caused by this PR or just required a rebase?

**Comment by Inve1951** [comment]
I don't quite understand the issue regarding the `mouse_entered`/`mouse_exited` signals. They fire correctly on 4.0.2 when dragging a child control outside its parent. But maybe I misunderstand - is it about firing them on the child itself?

What does not work in 4.0.2 however is the *mouse_up* event. It fires at the control where the user initiated the drag operation, and flags the event as handled, supressing both `_unhandled_input` and `_gui_input` at the cursor location - effectively making drag&drop behaviors impossible.
**Does this PR address that?**

See https://github.com/godotengine/godot/issues/20881#issuecomment-867027437 and https://github.com/godotengine/godot/issues/20881#issuecomment-883817011.

#54565 adds the boolean flag `Viewport.gui_force_mouse_events` which supposedly fixes that without a breaking change (while I personally think it should be the new default). But https://github.com/godotengine/godot/pull/54565#issuecomment-1427029304 suggests that #54565 is superseded by this PR (#67791).

**Comment by Sauermann** [comment]
@Inve1951 

> I don't quite understand the issue regarding the `mouse_entered`/`mouse_exited` signals. They fire correctly on 4.0.2 when dragging a child control outside its parent. But maybe I misunderstand - is it about firing them on the child itself?

Indeed, the issue is that a Control previously didn't receive `mouse_entered`/`mouse_exited`, while a Control-node is dragged. 

> What does not work in 4.0.2 however is the _mouse_up_ event. It fires at the control where the user initiated the drag operation, and flags the event as handled, supressing both `_unhandled_input` and `_gui_input` at the cursor location - effectively making drag&drop behaviors impossible. **Does this PR address that?**

No, this PR doesn't take care of that issue. However I aim to solve that issue with #67531.

> #54565 adds the boolean flag `Viewport.gui_force_mouse_events` which supposedly fixes that without a breaking change (while I personally think it should be the new default). But [#54565 (comment)](https://github.com/godotengine/godot/pull/54565#issuecomment-1427029304) suggests that #54565 is superseded by this PR (#67791).

If I remember correctly, #54565 wouldn't solve the problem of missing `mouse_entered`/`mouse_exited` events in other Windows / SubViewports.

**Comment by Sauermann** [comment]
> Seems to break popup `y` positions on macOS (but it might be unrelated and just need a rebase, I'll retest later). Couldn't find any other bugs, mouse enter/exit seems to work ok.

@bruvzg may I ask you to perform the retest on macOS? I am not aware of any Node-positions changes by this PR.

**Comment by Sauermann** [comment]
Please move to 4.2

**Comment by KoBeWi** [comment]
#57349 is not fixed by this PR.

**Comment by Sauermann** [comment]
As we discussed yesterday in #20881, mouse-entered/exited notifications now work (within a single Viewport).

#57349 states as use-case: "So I want to have a generic `is_over_ui` boolean to find out if my mouse is over UI or not."
This use-case can now be accomplished in GDscript by setting `is_over_ui` based on mouse-entered/exited notifications.

So I agree, that this PR doesn't solve the use-case of #57349 (in the scenario of a single viewport), because it already works. I will update the OP accordingly.

**Review by KoBeWi — APPROVED** [review]
I reviewed the code and it seems fine (not like I know much about it though) and the behavior is correct.
The issue I mentioned should be unlinked, unless you manage to fix it.

**Comment by Sauermann** [comment]
Thanks for the review. I have unlinked the mentioned issue.

**Comment by YuriSizov** [comment]
@RandomShaper Could you take a look here?

File: doc/classes/Control.xml
Comment (RandomShaper) [inline doc/classes/Control.xml:1141]: I have to admit I need to use the Eng-Spa dictionary to understand 'concealed.' Could you use a moer common word?

File: scene/main/viewport.cpp
Comment (RandomShaper) [inline scene/main/viewport.cpp:2942]: This reminds me of something we discussed in another PR, that it felt awkard to check the type of `this` that way. However, I don't remember where that was nor the decision. Do you know what I'm talking about?

**Review by RandomShaper — COMMENTED** [review]
What can I say? This is extremely complex. As others have said, code looks OK. Also, this has been reasonably tested, so I'd give it greenlight with a big TIWAGOS to let it be more exhaustively tested at this early point of the release cycle. I don't think there's a better way of proceeding unless we want to risk this rotting and we losing the benefits that it can provide.

File: doc/classes/Control.xml
Comment (lmurray) [reply doc/classes/Control.xml:1141]: "Occluded" is the de-facto technical term for this.

File: scene/main/viewport.cpp
Comment (Sauermann) [reply scene/main/viewport.cpp:2942]: Yes I remember, that was https://github.com/godotengine/godot/pull/77923#discussion_r1222519549
I will look into making this less awkward.

File: doc/classes/Control.xml
Comment (kleonc) [reply doc/classes/Control.xml:1141]: Is this statement true though? AFAICT after skimming through the code the sorting based on which these mouse notifcations occur doesn't take z-index into account. Which would mean it's not always a matter of the actual visible areas. :thinking:
https://github.com/godotengine/godot/blob/c3769e46b8a82640e14fbc8fca029787af757536/scene/main/viewport.cpp#L1623-L1693
https://github.com/godotengine/godot/blob/c3769e46b8a82640e14fbc8fca029787af757536/scene/main/viewport.cpp#L1381-L1389
https://github.com/godotengine/godot/blob/da81ca62a5f6d615516929896caa0b6b09ceccfc/scene/gui/control.h#L152-L160
https://github.com/godotengine/godot/blob/da81ca62a5f6d615516929896caa0b6b09ceccfc/scene/main/node.cpp#L1759-L1813

Comment (Sauermann) [reply doc/classes/Control.xml]: Thank you for pointing out the z-index. I did miss that.

Now I have been trying to describe the behavior in a way that includes all details but coming up with a concise way to do that is really hard. I hope, the current version is acceptable. I will gladly accept suggestions how to describe this in a better way.

File: scene/main/viewport.cpp
Comment (Sauermann) [reply scene/main/viewport.cpp]: I have implemented this in a different way.

File: scene/main/window.cpp
Comment (RandomShaper) [inline scene/main/window.cpp:683]: If that should never happen, as in "this is an engine bug," maybe a `DEV_ASSERT` is better (remember the condition has to be reversed for that macro).

**Review by RandomShaper — APPROVED** [review]
Let's put this in the wild!

File: scene/main/window.cpp
Comment (Sauermann) [reply scene/main/window.cpp]: Indeed this was intended for engine bugs. I agree, that `DEV_ASSERT` is a better fit.

**Comment by lawnjelly** [comment]
I wrote this on rocketchat earlier, as I was not sure whether `DEV_ASSERT` should have been used in this case (I originally wrote the macro, and Juan did warn of this potential problem at the time):

> Sauermann Just to note, I don't know whether DEV_ASSERT was needed there originally versus just printing an error and doing nothing (i.e. would it crash immediately afterwards)?
> In general we should prefer things like ERR_FAIL wherever possible because we don't want the engine to crash when there are bugs, and DEV_ASSERT will not act outside DEV_ENABLED builds.
> It should also be for performance critical stuff, there's no reason why there shouldn't be error checking in release builds here I think.
> I'll make a PR making note of this.

**Comment by 1xch** [comment]
I just updated to v4.2.dev3.official [013e8e3af] from v4.1.1 and I'm having a problem with what I suspect is this in particular.
Untitled
With the image above, the left is the window, the right is a snapshot of the node structure.  G_Window & G_Pane have widgets in them for handling capture, that previously just listened to mouse enter & exit. For example:

```
func _on_ready():
	_home.mouse_entered.connect(func(): LXN_FOCUSED.emit(true))
	_home.mouse_exited.connect(func(): LXN_FOCUSED.emit(false))
``` 

Now they do not work at all. Can I get a quick glance at this and maybe someone say why it needs to be this way when it did work. I'd like to know if this might be a bug, and I need to provide more info or documentation or if I need to rewrite how I'm handling capture on this. I'd really rather not as it was more a pain to poll mouse events before I just settled on using mouse enter and exit as the main part of measuring capture.

**Comment by Sauermann** [comment]
@1xch from your short description, it is difficult to guess, what the problem might be.
Also I'm not entirely sure, what you mean by "capture". In your code you emit a signal called "LXN_FOCUSED". Is the problem with mouse-capture or with focus?
If you could provide a minimal reproduction project of the problem, that you are experiencing, I could have a deeper look at it.

I just noticed, that you use a `SubViewport`, that is not a child of a `SubViewportContainer`. Have you implemented a way for sending InputEvents to that `SubViewport`?

**Comment by 1xch** [comment]
@Sauermann,
1. What I am saying is that the above worked. I've had it working for a while, with continual tinkering through most 4.0 versions, the above has been working for most of this year, with recent refactors this summer.
2. The problem is with mouse_entered & mouse_exited signals is what I've traced: they no longer happen. I would mostly ignore my terminology, I posted the above after discovery. 'Capture' here means acquiring mouse or other activity indicating the Control needs focused.
3. Yes, I've implemented a way for InputEvents to go to the SubViewport. That is not the issue.

I think a clearer statement is that G_Window & G_Pane in the above are controls that previously responded with mouse_entered & mouse_exited signals and now they do not. Both nodes still receive InputEvents, my core issue being the signals and not knowing why they aren't triggering when

**Comment by Sauermann** [comment]
@1xch Thanks for the MRP. I have tested it in v4.0.stable.official [92bee43ad], v4.1.1.stable.official [bd6af8e0e] and in v4.2.dev.custom_build [fa3428ff2] on Linux.

Your MRP behaved identical in my tests: when moving the mouse from `G_Window` to `G_Pane`, `G_Window` did receive a mouse-exited signal. So I can't replicate the overlapping behavior, that you describe. Please verify, if the MRP works for you as you would expect in v4.1.1.

I have tested the behavior in 3.5.1 and also there the parent Control receives a mouse-exited signal, when the mouse moves to a child-Control.

I would also kindly ask you to make future MRP less complicated: You use multiple `instantiantiate()` calls and different source-code files, which makes the project difficult and time-consuming to understand.

**Comment by wouldntsavezion** [comment]
Has there been a discussion about this somewhere? I see the work started back in 2022. I personally don't understand why the `mouse_entered` and `mouse_exited` signals behavior was changed. Also, this seems to be the source of many issues now, as well as effectively making `MOUSE_FILTER_PASS` largely unusable. 

Props to @Sauermann for all the work and now adding PRs to notify users in the docs, but this seems to me like it's just letting users know that events will not behave correctly. I think more discussion should definitely happen in #82530 , which sums up the situation quite well.

**Comment by emirljuca** [comment]
Yeah I gotta say, this broke my game too, I was relying on mouse pass to pass the mouse signal down while still being able to use the mouse entered and mouse exited signal. I have a control node that is further down the node stack that uses the mouse signal for a tooltip that will no longer work. Why make these changes? Is there any plans to add mouse pass work?

**Comment by kitbdev** [comment]
@emirljuca That issue should already be resolved in master as of #84547.

**Comment by emirljuca** [comment]
> @emirljuca That issue should already be resolved in master as of #84547.

That is great to hear, testing on newest build seems to show this to be the case, thanks!

**Pull request merged**
